### DISCUSSION TITLE
create controls but use it later?

### DISCUSSION THREAD

**Discussion by retsyo** [zauberzeug/nicegui#2173] [Q&A] [answered]
### Question

# 0. No, I don't mean to set visibility. 

# 1. story in short
As we know, if we write

```python
ui.label()
ui.element('video')
ui.button()
```

then the controls appear on the page in the order of `label` , `element` and `button` 

but what I need is

```python
ui.label()
ctrls4LaterUsage = ui.element('video').do_not_show_here() # it does not show on web page here
ui.button()
ctrls4LaterUsage.show() # but shows here
```

# 2. why not just create control later?

probably, you will suggest to create controls to follow this sequence: 

```python
ui.label()
ui.button()
ctrls4LaterUsage = ui.element('video')
```

but the JavaScript needs to read information from `ctrls4LaterUsage`, just like what falkoschindler has pointed out in ["capture user image?"](https://github.com/zauberzeug/nicegui/discussions/2010)

```python
video = ui.element('video').classes('w-[400px]')
canvas = ui.element('canvas')
ui.add_body_html(f'''
    <script type="text/javascript">
    addEventListener("DOMContentLoaded", (event) => {{
        window.webcam = new Webcam(getElement({video.id}), 'user', getElement({canvas.id}));
        window.webcam.start()
    }});
    </script>
''')
```

however in my code, later `add_body_html()` does not run due to [limitation of add_body_html](https://github.com/zauberzeug/nicegui/discussions/2140). Currently, I have to add all `add_body_html()` function calls at the very beginning of my function. So, the paradox occurs: I have to write`add_body_html()` at the beginning, which will use controls which are created later. If we can declare a controls but show it later, problem solves

```python
# a desired code

video = ui.element('video').classes('w-[400px]').create_but_donnot_show_here()
canvas = ui.element('canvas').create_but_donnot_show_here()
ui.add_body_html(f'''
    <script type="text/javascript">
    addEventListener("DOMContentLoaded", (event) => {{
        window.webcam = new Webcam(getElement({video.id}), 'user', getElement({canvas.id}));
        window.webcam.start()
    }});
    </script>
''')


ui.lable()
... # a lot of other controls/elements

video.show_here()
canvas.show_here()
```

Thanks

**Comment by falkoschindler** [comment, marked as answer]
A possible solution might be to _move_ the elements to show them somewhere else:
```py
# create elements
label_a = ui.label('A')
label_b = ui.label('B')

# use elements
print(label_a.id, label_b.id)

# place them somewhere else
container = ui.element()
label_a.move(container)
label_b.move(container)
```
